### PR TITLE
(feature/likebutton) OBJ3: BRIDGING THE ENLIGHENTED BACKEND WITH THE CURSED FRONTEND LANGUAGE KNOWN AS JAVASCRIPT. THIS IS MY PLEA TO USE REACT. I BEG OF YOU.

### LINKED ISSUE #21: front end functionality of chat feature

### LINKED ISSUE #22: back end functionality of chat feature w/ images

### LINKED ISSUE #23: Add chat interactivity

### PR DESCRIPTION
**Pull request by sumanyuj**
### Note: please merge and squash this AFTER Gia's, so Gia gets credit. A good portion of the backend and frontend is their stuff.

Before my changes here, you would not be able to have any posts, simply b/c the javascript and some of the backend functions were not compatible. 

### WHAT YOU CAN (and cannot) DO!

1) You must upload an image to get a chat. (this should be fixed in a bugfix) You can have a caption here. 

----> images do not work

2) You can like a post

2.5) You can do the above things as Guests (this is a feature, not a bug)

### TO DO:
 1) get images working (but not relevant for current OBJs)

2) get a scroll bar in the chat posts feed-box with a height in css

3) instead of hardcoding guest@buffalo.edu for each post, actually use DB in views.py

### CODE REVIEW

File: evergreen/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+asgiref==3.8.1 ; python_version >= "3.10" and python_version < "4.0"
```
Comment (Eric-Butcher) [inline evergreen/requirements.txt]: Delete all of this

**Review by Eric-Butcher — COMMENTED** [review]
We do not need to include migrations files.

**Review by Eric-Butcher — COMMENTED** [review]
Add a .gitignore to the migrations directory

File: evergreen/eg_app/models.py [hunk 2]
```diff
@@ -1,3 +1,27 @@
...
+    # user = models.ForeignKey(User, on_delete=models.CASCADE)
+    user = models.TextField(blank=False)
+    image = models.ImageField(upload_to="posts",blank=False,null=False)
+    caption = models.TextField(blank=True)
+    timestamp = models.DateTimeField(auto_now_add=True)
+    userLikes = models.ManyToManyField(User,blank=True)
+    likes = models.PositiveIntegerField(default=0)
+
+    def __str__(self):
+        # return f"{self.user.email}: {self.caption}"
```
Comment (Eric-Butcher) [inline evergreen/eg_app/models.py]: I think this `__str__` needs to be fixed

File: evergreen/eg_app/views.py [hunk 3]
```diff
@@ -35,4 +37,101 @@ def validate(request):
         return JsonResponse({"valid_pass":str(valid_pass),"valid_email":str(valid_email)})
 
 
+@csrf_exempt
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py:196]: Add some type hints this is confusing

File: evergreen/eg_app/views.py [hunk 4]
```diff
@@ -35,4 +37,101 @@ def validate(request):
...
+        }
+        posts_data.append(post_dict)
+
+    return JsonResponse({'posts': posts_data})
+@csrf_exempt
+def uploadPost(request):
+    if request.method == 'POST':
+
+        # user = request.user.email
+        user = 'guest@buffalo.edu'
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: Why is this always set to a constant string, and not using the version commented out above?

File: evergreen/eg_app/views.py [hunk 5]
```diff
@@ -35,4 +37,101 @@ def validate(request):
...
+            return JsonResponse({'status': 'success','image_url':post.image.url,'post_id': post.id})
+        else:
+            return JsonResponse({'status': 'error', 'message': 'Missing image or caption'}, status=400)
+    else:
+        return JsonResponse({'status': 'error', 'message': 'Invalid request method'}, status=405)
+
+
+
+
+@csrf_exempt
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py:240]: Add some comments saying this is not implemented yet, add `# TODO`.

File: evergreen/eg_app/views.py [hunk 6]
```diff
@@ -35,4 +37,101 @@ def validate(request):
...
+    if post.userLikes.contains(request.user):
+        post.likes -= 1
+        post.userLikes.remove(request.user)
+        post.save()
+    return redirect("/")
+
+@csrf_exempt
+def addComment(request, postId):
+    post = Post.objects.get(id=postId)
+    # user = request.user.email
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: Why is this hardcoded to guest?

File: evergreen/evergreen/urls.py [hunk 7]
```diff
@@ -21,13 +21,26 @@
...
     path("admin/", admin.site.urls),
     path('',views.index,name='index'),
     path('validate',views.validate,name='validate'), # type: ignore
 
+    path('validate',views.validate,name='validate'), # type: ignore
 
+    path('updateFeed',views.updateFeed,name="updateFeed"),
+    path('uploadPost',views.uploadPost,name="uploadPost"),
+    path('deletePost',views.deletePost,name="deletePost"),
+    path('likePost/<uuid:pk>', views.likePost, name="likePost"),
```
Comment (Eric-Butcher) [inline evergreen/evergreen/urls.py:35]: This is giving me squiggles.

File: evergreen/manage.py [hunk 8]
```diff
@@ -17,6 +17,16 @@ def main():
         ) from exc
     execute_from_command_line(sys.argv)
 
+if __name__ == "__main__":
+    main()
+
+#!/usr/bin/env python
+"""Django's command-line utility for administrative tasks."""
```
Comment (Eric-Butcher) [inline evergreen/manage.py]: Why does thie occur again? This lloks like a copy paste error.

File: evergreen/requirements.txt (hunk 1 above)
Comment (sumanyuj) [reply evergreen/requirements.txt]: Done

File: README.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 # Project Evergreen
 
-Project Evergreen is a group project for the class CSE312: Web Applications at the University at Buffalo in the Fall 2024 semester. This project is intended to become the web application for the UB IEEE student chapter automated greenhouse chapter. 
+Project Evergreen is a group project for the class CSE312: Web Applications at the University at Buffalo in the Fall 2024 semester. This project is intended to become the web application for the UB IEEE student chapter automated greenhouse chapter.
```
Comment (Eric-Butcher) [inline README.md:3]: automated greenhouse project not chapter

This might be better completed in another PR.

File: evergreen/Dockerfile [hunk 10]
```diff
@@ -21,9 +21,8 @@ RUN chmod +x /wait
 RUN ["/wait"]
 
 SHELL [ "poetry", "run" ]
+CMD [ "python3","manage.py","runserver","0.0.0.0:8080"]
```
Comment (WeirdAlex03) [inline evergreen/Dockerfile]: idk why this was moved but the file is outdated anyway. just bring this back down to remove this file from the PR and avoid a conflict

File: evergreen/manage.py [hunk 11]
```diff
@@ -1,9 +1,9 @@
+
```
Comment (WeirdAlex03) [inline evergreen/manage.py]: Delete this newline. That actually breaks functionality (the command to tell it to run as python MUST be the first line, otherwise this will get treated as a shell script)

File: evergreen/manage.py [hunk 12]
```diff
@@ -17,6 +17,16 @@ def main():
...
+if __name__ == "__main__":
+    main()
+
+#!/usr/bin/env python
+"""Django's command-line utility for administrative tasks."""
+import os
+import sys
+
+
+
```
Comment (WeirdAlex03) [inline evergreen/manage.py]: ???

I don't know how you pulled this off. but please un-pull this off.

File: evergreen/poetry.lock [hunk 13]
```diff
@@ -77,31 +77,44 @@ tzdata = {version = "*", markers = "sys_platform == \"win32\""}
...
 wmi = ["wmi (>=1.5.1)"]
 
 [[package]]
 name = "email-validator"
 version = "2.2.0"
 description = "A robust email address syntax and deliverability validation library."
+<<<<<<< HEAD
+=======
 
+>>>>>>> dev
```
Comment (WeirdAlex03) [inline evergreen/poetry.lock]: delete this

File: README.md (hunk 9 above)
Comment (WeirdAlex03) [inline README.md:3]: Removing the trailing spaces isn't bad, but I'd un-do it anyway just to avoid the conflict with Eric's rewriting of the README on merge

File: evergreen/poetry.lock (hunk 13 above)
Comment (sumanyuj) [reply evergreen/poetry.lock]: done!

File: evergreen/manage.py (hunk 12 above)
Comment (sumanyuj) [reply evergreen/manage.py]: done

File: README.md (hunk 9 above)
Comment (sumanyuj) [reply README.md:3]: done

File: evergreen/manage.py (hunk 11 above)
Comment (sumanyuj) [reply evergreen/manage.py]: done

**Review by Eric-Butcher — CHANGES_REQUESTED** [review]
__init__.py should not have been deleted

File: evergreen/eg_app/util/encrypt.py [hunk 14]
```diff
@@ -1,10 +1,11 @@
 import bcrypt
 
+
 def encrypt(plaintext):
 
     # encode plaintext string to bytes
     plaintext = plaintext.encode('')
```
Comment (Eric-Butcher) [inline evergreen/eg_app/util/encrypt.py]: just use .encode() or .encode(encoding="utf-8")

File: evergreen/eg_app/views.py [hunk 15]
```diff
@@ -217,6 +201,69 @@ def login_view(request: HttpRequest):
     return HttpResponseBadRequest()
 
 @csrf_exempt
+
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py]: whiteline between annotation and function defintion

File: evergreen/eg_app/views.py [hunk 16]
```diff
@@ -217,6 +201,69 @@ def login_view(request: HttpRequest):
...
+            'id': post.id,
+            'user': post.user,
+            'image': {'url': post.image.url if post.image else ''},
+            'caption': post.caption,
+            'likes': post.likes,
+            'comments': []
+        }
+        posts_data.append(post_dict)
+
+    return JsonResponse({'posts': posts_data})
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py:212]: add a newline

File: evergreen/eg_app/views.py (hunk 6 above)
Comment (sumanyuj) [reply evergreen/eg_app/views.py]: We will need to change that.

File: evergreen/eg_app/views.py [hunk 17]
```diff
@@ -217,6 +201,69 @@ def login_view(request: HttpRequest):
     return HttpResponseBadRequest()
 
 @csrf_exempt
+
+def updateFeed(request):
+
+    posts = Post.objects.all().order_by('-timestamp')
+    posts_data = []
```
Comment (Eric-Butcher) [inline evergreen/eg_app/views.py:200]: type annotation

File: evergreen/staticfiles/js/functions.js [hunk 18]
```diff
@@ -1,47 +1,88 @@
...
-    request.send(JSON.stringify(postJSON));
+  const form = document.getElementById("uploadForm");
+  const formData = new FormData(form);
+
+  // Add the frickinCSRF token
+  const csrfToken = document.querySelector("[name=csrfmiddlewaretoken]").value;
+
+  const request = new XMLHttpRequest();
+  request.open("POST", "uploadPost");
+  request.setRequestHeader("X-CSRFToken", csrfToken);
```
Comment (Eric-Butcher) [inline evergreen/staticfiles/js/functions.js:62]: Is this the correct way to name?

File: evergreen/staticfiles/js/functions.js [hunk 19]
```diff
@@ -1,47 +1,88 @@
...
+
+  request.onload = function () {
```
Comment (Eric-Butcher) [inline evergreen/staticfiles/js/functions.js:64]: We don't need this console.log I believe unless it is here for a reason

File: .gitignore [hunk 20]
```diff
@@ -82,6 +82,9 @@ target/
 profile_default/
 ipython_config.py
 
+# DJANGO Migrations
+migrations/
```
Comment (Eric-Butcher) [inline .gitignore]: You did this wrong I believe. We need `__init__.py` inside migrations. You should have seperate .gitignore inside the migrations folder configured to ignore everything except `__init__.py`.

**Review by Eric-Butcher — CHANGES_REQUESTED** [review]

File: README.md [hunk 21]
```diff
@@ -9,15 +9,23 @@ The development environment is being managed using [Poetry](https://python-poetr
...
 
-The virtual environment can be created locally by using the `poetry install` command from the `evergreen` directory. The environment can then be activated using the `poetry shell` command. This venv can then be deactivated by using the `exit` command. 
+The virtual environment can be created locally by using the `poetry install` command from the `evergreen` directory. The environment can then be activated using the `poetry shell` command. This venv can then be deactivated by using the `exit` command.
 
-New package dependencies for the project can be added using the `poetry add <package-name-from-pip>` command. 
+New package dependencies for the project can be added using the `poetry add <package-name-from-pip>` command.
 
 
 ## Setting up the Docker Environment
 In the root directory of the project, type `docker compose up --build --force-recreate` to spin up the containers. The poetry environment does not need to be activated for this.
```
Comment (WeirdAlex03) [inline README.md:18]: You made a mess out of this section too, lotsa duplicated paragraphs. Revert this to just Eric's verison

File: evergreen/staticfiles/js/functions.js [hunk 22]
```diff
@@ -1,61 +1,101 @@
...
-    request.open("DELETE", "deletePost");
-    request.send();
-}
+  const request = new XMLHttpRequest();
+  request.open("DELETE", "deletePost");
+  request.send();
+
 
-function deleteComment(commentId) {
+  function deleteComment(commentId) {
```
Comment (WeirdAlex03) [inline evergreen/staticfiles/js/functions.js]: this addition is all duplicated

File: evergreen/staticfiles/css/styles.css [hunk 23]
```diff
@@ -79,13 +83,17 @@ footer {
...
+
     height: 15%; 
+
 }
 
 .feed_posts {
     display: inline;
     align-content: center;
+
     padding: 5%
```
Comment (WeirdAlex03) [inline evergreen/staticfiles/css/styles.css]: you dropped the semicolon here. every other change in this file is unnecessary whitespace that i think was better before

File: evergreen/evergreen/urls.py [hunk 24]
```diff
@@ -24,19 +24,26 @@
 #from evergreen import settings
 
 urlpatterns = [
+
```
Comment (WeirdAlex03) [inline evergreen/evergreen/urls.py:27]: take out the type: ignore, those errors are there for a reason

**Review by WeirdAlex03 — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 25]
```diff
@@ -82,6 +82,8 @@ target/
 profile_default/
 ipython_config.py
 
+
+
```
Comment (WeirdAlex03) [inline .gitignore]: you left two newlines in here

**Review by WeirdAlex03 — CHANGES_REQUESTED** [review]

File: evergreen/staticfiles/js/functions.js (hunk 18 above)
Comment (sumanyuj) [reply evergreen/staticfiles/js/functions.js:62]: yes

File: .gitignore (hunk 20 above)
Comment (sumanyuj) [reply .gitignore]: done

**Review by WeirdAlex03 — APPROVED** [review]
:+1:
image

**Review by Eric-Butcher — APPROVED** [review]

**Review by Eric-Butcher — APPROVED** [review]
